### PR TITLE
Asyncio Interface

### PR DESCRIPTION
**Pull request by jamesba**
This adds an additional optional asyncio compatible interface to the library for use only in versions of python above 3.5. In doing so it replicates the existing file-like interface for the most part, but many of the methods become coroutines instead, and for any method that returns a list the asynchronous equivalent is an asynchronous generator.

### CODE REVIEW

**Comment by jamesba** [comment]
@barneygale What do you think of this? There doesn't seem to be an existing standard for how to do seekable file-like io with asyncio yet so I've had to improvise somewhat.

**Comment by barneygale** [comment]
I've tried to get my head around python's asyncio a couple of times and found it completely baffling, so I'm not much use for a code review. Let me know if/when you're comfortable for this to be merged and I'll merge it.

**Comment by jamesba** [comment]
Honestly it's not well documented at all, and has been something of a moving target. Hopefully it's a bit more stable now, and better documentation and tutorials should be forthcoming.

My understanding is roughly thus:

- The key to the whole thing is runloops. A runloop is an object which can be instantiated normally, and has a callable `run_until_complete` which can be called to make it run. When a run-loop is running it takes up the current thread exactly like a normal python method call and only returns when all the tasks in the run loop have completed. You have to explicitly call a runloop and give it at least one task in order to make any asynchronous code run at all.
- A task is a thing currently assigned to a runloop. A single runloop can have multiple tasks or just one. Only one task is actually running at any time, and all the others are dormant. Whenever the current task goes dormant the runloop uses a scheduling algorithm to determine which of its tasks to wake up. If none can be woken up it waits and tries again when one can be.
- Each task has a code execution stack much like an ordinary thread. The difference is that the code in a task is asynchronous code, which means it can go dormant and wait for conditions before waking up again.
- Asynchronous code is written by defining coroutines using `async def`.
- Inside asynchronous code you can call ordinary synchronous code and it's handled normally. You can't call asynchronous code from synchronous code.
- Inside asynchronous code you can `await` another asynchronous coroutine. This essentially acts like a function call except that the code being called is asynchronous and so can become dormant during the call and can `await` other things itself.
- Inside asynchronous code you can `await` a task. This causes the current task to become dormant until that other task has completed.
- There are also other things you can `await`, like semaphores and conditions, and the time.
- `async with` and `async for` are like `with` and `for` except that the magic methods used for context management/iteration are asynchronous coroutines and so can contain asynchronous code.

So basically you can only use the `await`, `async with` and `async for` keywords inside a coroutine defined with `async def`. A coroutine can only be run inside a task inside a runloop. A runloop will schedule across multiple tasks each time the current one is suspended. A task will never be suspended inside ordinary synchronous code, but can be suspended inside any `await`, `async with` or `async for` statement.

That's how I understand it, anyway.

**Comment by philipnbbc** [comment]
I think [open](https://github.com/jamesba/httpio/blob/jamesba-asyncio/httpio/__init__.py#L18) should call `HTTPIOFile.open` so that the behaviour is the same as before and it does what a user expects, e.g. raise an exception at the point if it fails to connect.

**Comment by philipnbbc** [comment]
Should `httpio.async` be renamed to `httpio.asyncio` (as you've done elsewhere) to avoid clashes with a reserved name?

File: httpio_async/__init__.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+        self._kwargs = kwargs
+        self._cursor = 0
+        self._cache = {}
+        self._session = None
+        self._aiter = None
+
+        self.length = None
+        self.closed = True
+
```
Comment (philipnbbc) [inline httpio_async/__init__.py]: Should this class have a `__repr__` similar to SyncHTTPIOFile?

File: httpio/__init__.py [hunk 2]
```diff
@@ -31,35 +32,44 @@ class HTTPIOError(IOBaseError):
...
+            try:
+                self.length = int(response.headers['Content-Length'])
+            except KeyError:
+                raise HTTPIOError("Server does not report content length")
+            if response.headers.get('Accept-Ranges', '').lower() != 'bytes':
+                raise HTTPIOError("Server does not accept 'Range' headers")
 
     def close(self):
         if not self.closed:
             self._session.close()
```
Comment (philipnbbc) [inline httpio/__init__.py]: Need to check that `_session` and `_cache` below are not None. Creating a SyncHTTPIO and then closing it before any operations will raise an exception.

File: httpio_async/__init__.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+            self._aiter = self.readlines()
+
+        return await self._aiter.__anext__()
+
+    async def open(self):
+        """This method has no direct equivalent in the HTTPIOFile interface, but does things which
+        are the equivalent of what's done in the constructor of that class. Since constructors cannot
+        be coroutines this class needs this as a seperate coroutine"""
+
+        if self.closed:
```
Comment (philipnbbc) [inline httpio_async/__init__.py]: Should this check `_session` instead to be consistent with SyncHTTPIO, or SyncHTTPIO be changed? E.g. if `open` fails with an exception raised then a second attempt should do nothing to be consistent. I haven't checked when BufferedIOBase or IOBase set `closed` and that might be relevant here as well.

File: httpio_async/__init__.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+    async def close(self):
+        if not self.closed:
+            await self._session.__aexit__(None, None, None)
+            self._session = None
+            self._cache.clear()
+            self.closed = True
+
+    async def flush(self):
+        self._cache.clear()
```
Comment (philipnbbc) [inline httpio_async/__init__.py]: SyncHTTPIO has an `_assert_open()` here in flush. Perhaps that should be removed from there?

File: httpio_async/__init__.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+                    if b'\n' in sector:
+                        break
+                if b'\n' in sector:
+                        break
+
+        data = data.splitlines(True)[0]
+        self._cursor += len(data)
+        return data
+
+    async def readlines(self, size=-1):
```
Comment (philipnbbc) [inline httpio_async/__init__.py]: Parameter is called `hint` in IOBase.

File: httpio_async/__init__.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+        self._assert_open()
+        if whence == 0:
+            self._cursor = offset
+        elif whence == 1:
+            self._cursor += offset
+        elif whence == 2:
+            self._cursor = self.length + offset
+        else:
+            raise HTTPIOError("Invalid argument: whence=%r" % whence)
+        if not (0 <= self._cursor <= self.length):
```
Comment (philipnbbc) [inline httpio_async/__init__.py]: Not related to this PR but possibly interesting to note that a seek beyond the limits should be allowed without raising an exception. A file object from `open` allows it similar to C. I noticed it only because `_read_impl` has a `if size == 0` check, not <= 0.

File: tox.ini [hunk 7]
```diff
@@ -8,7 +8,8 @@ envlist = py27, py3
 
 [testenv]
 commands =
-    python -m unittest discover -s tests
+    py27: python -m unittest discover -s tests -p 'test_sync*.py'
+    py3: python -m unittest discover -s tests -p 'test*.py'
```
Comment (philipnbbc) [inline tox.ini]: The `py3` test would fail for a system that has python version < 3.6 (and maybe 3.7?). Should `py3` in the envlist be changed to `py36` as it would be clearer what is being tested and why it fails? You did something similar elsewhere and maybe the same should be done here. The testenv command could include `py35` as well in case a user wants to test that using the tox `-e` option.

File: httpio_async/__init__.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+
+    async def _read_cached(self, size, max_raw_reads=-1):
+        sector0, offset0 = divmod(self._cursor, self.block_size)
+        sector1, offset1 = divmod(self._cursor + size - 1, self.block_size)
+        offset1 += 1
+        sector1 += 1
+
+        raw_reads = 0
+
+        data = []
```
Comment (philipnbbc) [inline httpio_async/__init__.py]: Detected by flake8 - data variable is not used.

File: httpio_async/__init__.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+    The returned object is open, and needs to be closed when done.
+    This will be handled automatically if you use it as an asynchronous context manager.
+
+    :param url: The URL of the file to open
+    :param block_size: The cache block size, or `-1` to disable caching.
+    :param kwargs: Additional arguments to pass to `requests.Request()`
+    :return: An `httpio.HTTPIOFile` object supporting most of the usual
+        file-like object methods.
+    """
+    f = AsyncHTTPIOFile(session, url, block_size, **kwargs)
```
Comment (philipnbbc) [inline httpio_async/__init__.py]: Detected by flake8 - `session` is undefined.

**Comment by philipnbbc** [comment]
Could a `.flake8` file be added? It has exposed 1 real issue.

File: httpio_async/__init__.py (hunk 3 above)
Comment (jamesba) [reply httpio_async/__init__.py]: The base classes set closed to False upon initialisation, which is different behaviour from the Async version.

**Comment by jamesba** [comment]
@philipnbbc care to take another look? This should have fixed all your comments.

**Comment by philipnbbc** [comment]
@jamesba the tests seem to take an awful lot longer after the new commits, i.e. after 52db224121c2abaa8a3639ddc41f64593425cd28. I can't figure out what change could be causing that.

Running `tox -e py36` using the latest code now results in this:
```
GLOB sdist-make: /store1/external/jamesba-httpio/setup.py
py36 inst-nodeps: /store1/external/jamesba-httpio/.tox/dist/httpio-0.2.0.zip
py36 installed: aiohttp==3.5.4,async-timeout==3.0.1,attrs==19.1.0,certifi==2019.3.9,chardet==3.0.4,coverage==4.5.3,httpio==0.2.0,idna==2.8,idna-ssl==1.1.0,mock==2.0.0,multidict==4.5.2,pbr==5.1.3,requests==2.21.0,six==1.12.0,typing-extensions==3.7.2,urllib3==1.24.2,yarl==1.3.0
py36 runtests: PYTHONHASHSEED='1234660441'
py36 runtests: commands[0] | python -m unittest discover -s tests -p test*.py
Executing <Task finished coro=<TestAsyncHTTPIOFile.test_tell_starts_at_zero() done, defined at /store1/external/jamesba-httpio/tests/test_async_httpio.py:268> result=None created at /usr/lib/python3.7/asyncio/base_events.py:563> took 127.231 seconds
.Executing <Task finished coro=<test_readline() done, defined at /store1/external/jamesba-httpio/tests/test_async_httpio.py:247> result=None created at /usr/lib/python3.6/asyncio/base_events.py:452> took 127.231 seconds
.Executing <Task finished coro=<TestAsyncHTTPIOFile.test_throws_exception_when_get_returns_error() done, defined at /store1/external/jamesba-httpio/tests/test_async_httpio.py:192> result=None created at /usr/lib/python3.7/asyncio/base_events.py:563> took 127.231 seconds
.Executing <Task finished coro=<test_readlines() done, defined at /store1/external/jamesba-httpio/tests/test_async_httpio.py:254> result=None created at /usr/lib/python3.6/asyncio/base_events.py:452> took 127.230 seconds
.Executing <Task finished coro=<test_aiter() done, defined at /store1/external/jamesba-httpio/tests/test_async_httpio.py:261> result=None created at /usr/lib/python3.6/asyncio/base_events.py:452> took 127.304 seconds
...
```
at which point I cancelled it.

If I run it at commit 52db224121c2abaa8a3639ddc41f64593425cd28 it takes around 30 seconds.

Do you see the same thing?

**Comment by jamesba** [comment]
@philipnbbc I did, I thought it was a local problem with the network drive I was using. Right, time to try and do a git bisect to work out what has caused this.

**Comment by philipnbbc** [comment]
The changes LGTM however. The only thing I spotted that might need changing is `async def close` t check that self._session is not None for case where open was not called.

**Comment by jamesba** [comment]
Bisect identifies commit 9a3751b9e839da68361c378d7ad7abae71bc5c61 as having introduced the problem. I will take a look at cutting that commit up further on Monday.

**Comment by jamesba** [comment]
@philipnbbc I have identified and reverted the change that caused the timing problem. I will explore further to try and work out how it caused that, and push a version without the problem.

**Comment by jamesba** [comment]
@philipnbbc this should now be working

File: httpio_async/__init__.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+    async def __aenter__(self):
+        await self.open()
+        return self
+
+    async def __aexit__(self, exc_type, exc, tb):
+        await self.close()
+
+    async def close(self):
+        if not self.closed:
+            await self._session.__aexit__(None, None, None)
```
Comment (philipnbbc) [inline httpio_async/__init__.py]: Add a check here that _session is not None? This is required when open has not been called.

Comment (jamesba) [reply httpio_async/__init__.py]: If open has not been called then closed will be False.

Comment (philipnbbc) [reply httpio_async/__init__.py]: I might be missing something, but it's initialised to `False` at line 52.

Comment (jamesba) [reply httpio_async/__init__.py]: You're right. I changed it and didn't update this to match. Will fix.

**Comment by jamesba** [comment]
@philipnbbc @barneygale right, all changes made, history cleaned up. Ready to merge.

**Comment by philipnbbc** [comment]
@jamesba LGTM.

**Comment by barneygale** [comment]
Thanks both
